File: src/tests/Behavioral/LocalTypeAliasScope/LocalTypeAliasScopeSecond.go
// The CROSS-FILE half of the local-type-alias scope guard. A `global using` is scoped to the
// whole compilation, so declaring `testFnc` and `fileMaker` here — names this package's other
// file already declares, in two of its functions — is the second shape of the same CS1537. This
// is archive/tar's reader_test.go against its writer_test.go.
//
// A single-file test could not catch it: the emission is per-file, and only the compilation sees
// both.
package main

import "fmt"

// secondWriteOps declares the same two local names a third and fourth time, again over an
// unrelated member set.
func secondWriteOps() {
	type (
		readOp struct {
			cnt int
		}
		testFnc   any // readOp
		fileMaker any // readOp
	)

	tests := []testFnc{readOp{1}, readOp{2}}
	maker := fileMaker(readOp{9})

	sum := 0

	for _, t := range tests {
		if op, ok := t.(readOp); ok {
			sum += op.cnt
		}
	}

	if op, ok := maker.(readOp); ok {
		sum += op.cnt
	}

	fmt.Println("secondWriteOps:", len(tests), sum)
}

// secondAliases repeats the real-alias name too, and reaches the same package-level type the
// other file's alias targets.
func secondAliases() {
	type hdr = Header

	h := hdr{Name: "link.txt", Size: 0}

	fmt.Println("secondAliases:", h, Header{Name: "raw", Size: 7})
}

File: src/tests/Behavioral/LocalTypeAliasScope/LocalTypeAliasScopeSecond.cs
global using secondWriteOps_testFnc = object;
global using secondWriteOps_fileMaker = object;
global using secondAliases_hdr = go.main_package.Header;

namespace go;

using fmt = fmt_package;

partial class main_package {

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly object secondWriteOpsˢ = (@string)"secondWriteOps:"u8;

[GoType("dyn")] partial struct secondWriteOps_readOp {
    internal nint cnt;
}

internal static void secondWriteOps() {
    var tests = new secondWriteOps_testFnc[]{new secondWriteOps_readOp(1), new secondWriteOps_readOp(2)}.slice();
    var maker = ((secondWriteOps_fileMaker)new secondWriteOps_readOp(9));
    nint sum = 0;
    foreach (var (_, t) in tests) {
        {
            var (op, ok) = t._<secondWriteOps_readOp>(ᐧ); if (ok) {
                sum += op.cnt;
            }
        }
    }
    {
        var (op, ok) = maker._<secondWriteOps_readOp>(ᐧ); if (ok) {
            sum += op.cnt;
        }
    }
    fmt.Println(secondWriteOpsˢ, len(tests), sum);
}

// Hoisted @string literals (single allocation; Go keeps these in RODATA)
private static readonly object secondAliasesˢ = (@string)"secondAliases:"u8;

internal static void secondAliases() {
    var h = new secondAliases_hdr(Name: "link.txt"u8, Size: 0);
    fmt.Println(secondAliasesˢ, h, new Header(Name: "raw"u8, Size: 7));
}

} // end main_package
